src/models/marian.cpp

```diff
@@ -0,0 +1,244 @@
+  }
+
+  is_prompt_ = false;
+}
+
+DeviceSpan<float> MarianState::Run(int current_length, DeviceSpan<int32_t>& next_tokens, DeviceSpan<int32_t> next_indices) {
+  if (first_run_) {
+    // INITIALIZE THE ENCODER AND RUN IT ONCE
+
+    encoder_input_ids_.name_ = "input_ids";
+    encoder_input_ids_.Add();
+
+    encoder_attention_mask_.name_ = model_.config_->model.encoder.inputs.attention_mask;
+    encoder_attention_mask_.Add();
+
+    encoder_input_ids_.Update(next_tokens);
+    size_t new_length = static_cast<size_t>(encoder_input_ids_.GetShape()[1]);
+    encoder_attention_mask_.Update(next_tokens, current_length, static_cast<int>(new_length));
+
+    auto encoder_outputs_type = model_.session_info_.GetOutputDataType("encoder_outputs");
+    auto encoder_outputs_shape = std::array<int64_t, 3>{encoder_input_ids_.GetShape()[0], encoder_input_ids_.GetShape()[1], 512};
+    encoder_outputs_ = OrtValue::CreateTensor(model_.p_device_->GetAllocator(), encoder_outputs_shape, encoder_outputs_type);
+
+    output_names_.push_back("encoder_outputs");
+    outputs_.push_back(encoder_outputs_.get());
+
+    State::Run(*model_.session_encoder_);
+
+    // CLEAR INPUTS AND OUTPUTS
+    ClearIO();
+
+    // INITIALIZE THE DECODER
+    decoder_input_ids_.name_ = "input_ids";
+    decoder_input_ids_.AddMarianInputs();
+
+    next_tokens.CpuSpan()[next_tokens.size() - 1] = 32000;
+
+    decoder_input_ids_.Update(next_tokens.subspan(next_tokens.size() - 1, 1));
+
+    const std::array<int64_t, 1> past_key_values_length_shape{1};
+    past_key_values_length_ = OrtValue::CreateTensor(model_.allocator_cpu_, past_key_values_length_shape, model_.session_info_.GetInputDataType(model_.config_->model.decoder.inputs.past_key_values_length));
+    input_names_.push_back(model_.config_->model.decoder.inputs.past_key_values_length.c_str());
+    inputs_.push_back(past_key_values_length_.get());
+
+    *past_key_values_length_->GetTensorMutableData<int64_t>() = -1;
+
+    attention_mask_.name_ = model_.config_->model.decoder.inputs.encoder_attention_mask;
+    attention_mask_.Add();
+    attention_mask_.Update(next_tokens, current_length, static_cast<int>(new_length));
+
+    auto hidden_states_type = model_.session_info_.GetInputDataType("encoder_hidden_states");
+    auto encoder_hidden_states_shape = std::array<int64_t, 3>{decoder_input_ids_.GetMarianInputsShape()[0], encoder_input_ids_.GetShape()[1], 512};
+    encoder_hidden_states_ = OrtValue::CreateTensor(model_.p_device_->GetAllocator(), encoder_hidden_states_shape, hidden_states_type);
+
+    input_names_.push_back("encoder_hidden_states");
+    inputs_.push_back(encoder_outputs_.get());
+
+    auto rnn_states_prev_type = model_.session_info_.GetInputDataType("rnn_states_prev");
+    auto rnn_states_prev_shape = std::array<int64_t, 3>{3, decoder_input_ids_.GetMarianInputsShape()[0], 512};
+    rnn_states_prev_ = OrtValue::CreateTensor(model_.p_device_->GetAllocator(), rnn_states_prev_shape, rnn_states_prev_type);
```

Comment:
If the input is allocated on the device, it may not be CPU accessible. Please remove the for loop below and use the device interface to perform device array manipulations.